core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1202,8 +1202,9 @@ private JoinOnLeftCalcToJoinUnifyRule() {
       if (joinRelType == null) {
         return null;
       }
-      if (joinRelType != JoinRelType.INNER
-          && !(joinRelType.isOuterJoin() && qInput0Cond.isAlwaysTrue())) {
+      // If JoinType is inner-join or out-join,
+      // Calc which is bottom join can be pulled up.
+      if (joinRelType != JoinRelType.INNER && !joinRelType.isOuterJoin()) {
```

Comment:
What if it's a right Join but with Calc as the left child ?
If above concern is valid, please add a corresponding test.